### ISSUE TITLE
[FR] - Firebase Admin Lite

### ISSUE THREAD

**Issue by jdgamble555** [firebase/firebase-admin-node#2572]
**Is your feature request related to a problem? Please describe.**
Currently Firebase cannot run on Deno, Cloudflare, Vercel Edge, or Bun (among others). Firebase will mainly be used in Node Serverless environments with Cold Starts, which are unacceptable for many app developers.

**Describe the solution you'd like**
If you don't require Authentication, you can use the "Firestore Lite" package. However, we cannot use Authentication or the Server Functions in order to keep our session. This is a HUGE problem and limits where Firebase can be deployed.

**Describe alternatives you've considered**
Using the REST API. However, it is EXTREMELY cumbersome in order to convert JSON data to the data types that REST API uses, as well as the problems handling tokens etc.

**Additional context**
- https://github.com/firebase/firebase-admin-node/issues/1879 - This should have been reopened, but never was.
- https://firebase.uservoice.com/forums/948424-general/suggestions/48355772-firebase-admin-lite - User Voice Feature Request

It would be nice to be able to use the [FirebaseServerApp](https://firebase.blog/posts/2024/05/firebase-serverapp-ssr) as well, but this would not have ALL the features we may need in other JS environments.

J

**Comment by dconeybe** [comment]
@jdgamble555 Thank you for raising this issue. Could you elaborate on the use case where the `preferRest` option (as noted in https://github.com/firebase/firebase-admin-node/issues/1879) does not solve your issue?

**Comment by jdgamble555** [comment]
Yes, this will not solve the problem as "Edge" computing cannot load packages like "fs", "stream", "child_process", "crypto", "os", etc... This are usually items that are required with grpc,  protobuf protocol etc... but NOT with REST API.

J

**Comment by jdgamble555** [comment]
setting `preferRest` still loads these items

**Comment by dconeybe** [comment]
Ok I think I see what you're saying. Could you mention which Firebase products specifically would be useful to you in a "lite" SDK? You mentioned Firestore for one. Also, would you prefer a "tree-shakeable" API (like https://github.com/firebase/firebase-js-sdk) or an object-oriented API (like https://github.com/firebase/firebase-admin-node)?

In any case, this is a pretty big undertaking so I have no promises as to if, or when, such a product will become available. But it's helpful to gather yours (and others') requirements for a time when we decide to scope such a project.

**Comment by dconeybe** [comment]
Also, could you describe how you would envision authentication to work? Authentication in the firebase-node-admin SDK in non-google edge computing environments would likely need an "exported service account key file" (https://firebase.google.com/docs/admin/setup) and these environments often do not have access to the file system. Would it work, for example, to be able to specify the JSON contents of the key file rather than a _path_ to the key file? Then, at runtime, the code would need to first obtain the key file from some other place (e.g. d1/kv/r2 in cloudflare workers) and specify it to the initialization function for Firebase.

**Comment by jdgamble555** [comment]
Here is how I setup my [Firebase Admin](https://code.build/p/sveltekit-todo-app-with-firebase-admin-tqdc5j#firebase-admin) app now. I use the environment variables.

Firestore already has a "lite" version that [works with the REST API](https://firebase.google.com/docs/firestore/solutions/firestore-lite). It will work on non-node runtimes. However, it is only Firestore (not Firebase Auth etc) so you can't use an authentication token nor fetch data behind Firestore Rules. 

Here is an example of a package that tries to mimic what Firebase Admin should already do.

https://github.com/Moe03/firebase-admin-rest

I have seen enough of these packages built over the years, that they never keep up-to-date. This is how it should work out-of-the-box IMO.

Imagine:

```ts
import { getAuth } from 'firebase-admin/auth/lite';
import { getFirestore } from 'firebase-admin/firestore/lite';
```
Similar to how the frontend works.

J

**Comment by dconeybe** [comment]
@jdgamble555 Thank you for the explanation and the example from @Moe03. I've logged b/343978422 internally to track this feature request and will keep you apprised of developments. But TBH it's unlikely that this feature request will see much progress without more requests from the community. For anyone who would like to see this feature, please reply with a brief explanation of your use case to help us prioritize this!

**Comment by ganey** [comment]
Wow this is still an ongoing issue? (i got the notification from the previous one with GRPC slow init)

iirc last time i used the admin sdk, preferRest was firebase/firestore only, so making one call to do an admin function meant the request took about 8 seconds in cloud functions/aws lambda.

If your doing firebase auth with a backend like php (no firebase sdk), then you could be using a node based admin app to handle this, and loading the admin sdk was just always slow.

8 seconds for a user is just ridiculous, and the early solutions such as 'function keep alive / warmers' are really not dev friendly...

'oh hey our sdk is massive and takes a while to load, so spend more money with us keeping your code cached and it'll be fast'.


Sorry if i come across annoyed by this, but I had many issues with firebase over the years from bad/missing/stale docs, to issues like this just taking forever to get fixed.

**Comment by Rizzist** [comment]
hoping new firebase admin lite works on edge

**Comment by kulterryan** [comment]
how to use it with Bun?

**Comment by dconeybe** [comment]
Note that this repository specifically targets node (the name is `firebase-admin-node`). Providing a version that works portably across JavaScript runtimes (e.g. deno, bun, cloudflare workers) is a significant undertaking to create and maintain long-term and not something that is prioritized at the moment. That being said, please continue to voice your interest in such an SDK so that we can get a sense of the customer demand for this feature. Thanks everyone!

**Comment by jdgamble555** [comment]
@dconeybe - To be fair, it was built for Node before other Runtimes were popular and probably named to differentiate between go, python etc.

It would be advantageous to build a "lite" version, just like `firebase-js-sdk` has, as it could speed up **NodeJS** loading times as well. I don't think you would need to rebuild the whole package, just allow more tree-shaking and less dependency needs per package. If you imported from `firebase-admin/lite` for example, it would not import all the bulk package dependencies etc.

J

**Comment by dconeybe** [comment]
@jdgamble555 Thank you for your input. I don't disagree that there is value in providing a version that is decoupled from node.js. It's just a matter of prioritization.

**Comment by juniorforlife** [comment]
I'm desperate for a lightweight version of the Firebase Admin SDK. There have been numerous complaints about the painfully slow cold start caused by the Admin SDK dating back to 2020.

**Comment by richardrauser** [comment]
Another vote for this.. we would like to use Firestore in a Vercel edge function, and as previously stated, the Vercel edge runtime is missing many packages reuqired by Firebase Admin.

**Comment by jimmy0251** [comment]
Another vote for this. We would like to have some functions at edge that depend on Firebase admin SDK.

**Comment by GriffinJohnston** [comment]
+1

Edge functions offer really compelling benefits for certain applications, and are becoming increasingly popular as a result. The lack of an official SDK for Firestore in these environments is a big problem.

**Comment by danielthedifficult** [comment]
Here to +1 this. Our team has lost so much time dealing with the REST interface 'manually' as it is not exactly developer friendly ;)

**Comment by chenlevy** [comment]
+1

**Comment by deloreyj** [comment]
+1

**Comment by cungminh2710** [comment]
+1

**Comment by juniorforlife** [comment]
@dconeybe There are a lot of people requesting this feature. Can you please share if the team has any plan for it?

**Comment by ganey** [comment]
We're 6 years in to the issue on slow starts, this obviously isn't getting done any time soon 🤦

**Comment by juniorforlife** [comment]
I'm shocked that Firebase admin node doesn't support serverless even though they are widely used and have been around for years

**Comment by ganey** [comment]
> I'm shocked that Firebase admin node doesn't support serverless even though they are widely used and have been around for years

It runs fine in regular cloud functions, or on AWS Lambda etc, but the cold start time of the admin SDK is still awful. You can't rely on it as part of any custom signup or login flow.

**Comment by lahirumaramba** [comment]
> > I'm shocked that Firebase admin node doesn't support serverless even though they are widely used and have been around for years
> 
> It runs fine in regular cloud functions, or on AWS Lambda etc, but the cold start time of the admin SDK is still awful. You can't rely on it as part of any custom signup or login flow.

@ganey by "cold start time of the admin SDK" do you mean the Firestore API? If that's the case, do you still experience longer cold starts with the `preferRest` option enabled?

**Comment by ganey** [comment]
> > > I'm shocked that Firebase admin node doesn't support serverless even though they are widely used and have been around for years
> > 
> > It runs fine in regular cloud functions, or on AWS Lambda etc, but the cold start time of the admin SDK is still awful. You can't rely on it as part of any custom signup or login flow.
> 
> @ganey by "cold start time of the admin SDK" do you mean the Firestore API? If that's the case, do you still experience longer cold starts with the `preferRest` option enabled?

Yeah firestore is better with preferRest, has been since it was implemented as the GRPC is slow to setup. The admin SDK causes a slow down on fresh serverless functions

**Comment by juniorforlife** [comment]
> > I'm shocked that Firebase admin node doesn't support serverless even though they are widely used and have been around for years
> 
> It runs fine in regular cloud functions, or on AWS Lambda etc, but the cold start time of the admin SDK is still awful. You can't rely on it as part of any custom signup or login flow. 

Slow cold start not is slow on AWS lambda but also completely broken on Vercel and Cloudflare

**Comment by ganey** [comment]
Hey @dconeybe / @lahirumaramba any movement on this? The original slow SDK bug was raised back in 2020, a few days short of 5 years ago now. Firestore has progressed in this timethe `preferRest` option to skip the slow gprc issues on cold starts which is great, but everyone's still waiting for movement on the admin sdk, or at least some "lite" functionality to make it more useable in "serverless" environments.

**Comment by dconeybe** [comment]
No updates, unfortunately.

**Comment by Aaron-at-TBL** [comment]
I think this is a striking example of corporate software engineering approaches. Clearly Google has more than enough resources to build a fantastic Firestore SDK, and doing so would greatly benefit the Firebase product, yet leadership is necessary to get Firestore SDK to where it could/should be.

**Comment by jdgamble555** [comment]
@lahirumaramba @dconeybe - Cloudflare, specifically, has come a long way with the `node_compat`. It seems MongoDB and Mongoose didn't originally work, but now they do.

https://github.com/cloudflare/workerd/discussions/2721

Perhaps the easiest way to get this working is to remove SOME of the nodejs npm packages, but NOT all. This is what MongoDB JS did.

Fixing it for Cloudflare would solve MOST use cases.

J

**Comment by androve2k** [comment]
I ran into exactly this on Cloudflare Workers (gRPC transport isn't available in V8 isolates, and the REST API's typed value format is painful to hand-roll). Ended up writing a small package that wraps the REST API with the JS↔Firestore type conversion handled, JWT signing done via Web Crypto (crypto.subtle) instead of Node's crypto/jsonwebtoken. It's intentionally scoped — get/set-merge/simple queries only, no transactions or realtime listeners — but covers the common case people are hitting here: [firestore-rest-edge on npm](https://www.npmjs.com/package/firestore-rest-edge) / [GitHub](https://github.com/androve2k/firestore-rest-edge). Might save someone the same yak-shaving.

**Comment by jdgamble555** [comment]
@lahirumaramba @dconeybe - Could we not just strip out the few Node only packages that cause issues?

Surely AI could find the culprits and this could be done easily these days?

Firebase Admin is the ONLY database NPM package that doesn't work on Cloudflare with `node-compat`.

MongoDB works now, full Postgres connection support, etc...

There are only a few incompatible packages that cause the issues...

We can write a custom new package, but Firebase has so many new features now that it would be impossible...

J
